modules/utils/graph/Redif.cpp

```diff
@@ -0,0 +1,325 @@
+			{
+				tempIdx[j] = j;
+				tempDist[j] = dist(j, i);
+			}
+			std::cout << i << "-th" << std::endl;
+			Quicksort(tempIdx, tempDist, 0, n - 1);
+			//if (i == 58)
+			//{
+			//	std::cout << std::endl;
+			//	for (size_t j = 0; j < 30; j++)
+			//		std::cout << tempDist[j] << std::endl;
+			//}
+			for (size_t j = 0; j < nNeighbors; j++)
+			{
+				knnMat(j, i) = tempIdx[j + 1];
+				knnDistMat(j, i) = tempDist[j + 1];
+			}
+		}
+
+		blaze::DynamicMatrix<Tv> localDist(n, n);
+
+		for (size_t i = 0; i < nNeighbors; i++)
+			for (size_t j = 0; j < n; j++)
+				localDist(ixx(i, j), knnMat(i, j)) = knnDistMat(i, j);
+
+		blaze::DynamicMatrix<Tv> retLocalDist(n, n);
+		for (size_t i = 0; i < n; i++)
+			for (size_t j = 0; j < n; j++)
+				retLocalDist(i, j) = (localDist(i, j) > localDist(j, i)) ? localDist(i, j) : localDist(j, i);
+
+		return retLocalDist;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::calcWeightedGraphLaplacian(blaze::DynamicMatrix<Tv> localDist)
+	{
+		size_t n = localDist.rows();
+
+		/// Degree matrix
+		blaze::DynamicMatrix<Tv> D(n, n);
+		for (size_t i = 0; i < D.rows(); i++)
+		{
+			double sum = 0;
+			for (size_t j = 0; j < n; j++)
+				sum += localDist(i, j);
+			if (sum == 0) sum = 1;
+			D(i, i) = 1 / sum;
+		}
+
+		blaze::DynamicMatrix<Tv> A(localDist.rows(), localDist.columns());
+		A = D * localDist * D / n;
+
+		blaze::DynamicMatrix<Tv> D1(n, n);
+		for (size_t i = 0; i < D.rows(); i++)
+		{
+			double sum = 0;
+			for (size_t j = 0; j < n; j++)
+				sum += A(i, j);
+			if (sum == 0) sum = 1 / n;
+			D1(i, i) = sum;
+		}
+
+		blaze::DynamicMatrix<Tv> graphLap(n, n);
+		graphLap = D1 - A;
+
+		return graphLap;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::encode(blaze::DynamicMatrix<Tv> x, blaze::DynamicVector<Tv> & l_idx)
+	{
+		size_t nTrain = xTrain.rows();
+		size_t nX = x.rows();
+		size_t nIter = LArray.size();
+		/// Find nearest Neighbors of each record of x in xTrain
+		size_t l = 1; /// Number of nearest neighbors
+		l_idx = blaze::DynamicVector<Tv>(nX);
+		unsigned size_t k_check = (unsigned size_t)sqrt(nTrain); /// Parameter for checking model on the record
+
+		/// Compute distances
+		for (size_t i = 0; i < nX; i++)
+		{
+			blaze::DynamicMatrix<Tv> distTotal(nTrain + 1, nTrain + 1);
+			for (size_t k = 0; k < nTrain + 1; k++)
+				for (size_t j = 0; j < nTrain + 1; j++)
+				{
+					if (k == j) continue;
+
+					blaze::DynamicVector<Tv, rowVector> rowk;
+					blaze::DynamicVector<Tv, rowVector> rowj;
+
+					if (k < nTrain) rowk = row(xTrain, k);
+					else rowk = row(x, i);
+					if (j < nTrain) rowj = row(xTrain, j);
+					else rowj = row(x, i);
+
+					std::vector<Tv> veck, vecj;
+					for (size_t s = 0; s < rowk.size(); ++s)
+					{
+						veck.push_back(rowk[s]);
+						vecj.push_back(rowj[s]);
+					}
+
+					Metric metric;

```

Comment:
this code makes it really hard to provide custom metric. can we make it a method's argument instead?